core/src/main/scala/kafka/controller/TopicDeletionManager.scala

```diff
@@ -62,7 +62,27 @@ class TopicDeletionManager(controller: KafkaController,
   val controllerContext = controller.controllerContext
   val isDeleteTopicEnabled = controller.config.deleteTopicEnable
   val topicsToBeDeleted = mutable.Set.empty[String]
-  val partitionsBeingDeleted = mutable.Set.empty[TopicPartition]
+  /** The following topicsBeingDeleted variable is used to properly update the offlinePartitionCount metric.
+    * When a topic is going through deletion, we don't want to keep track of its partition state
+    * changes in the offlinePartitionCount metric, see the PartitionStateMachine#updateControllerMetrics
+    * for detailed logic. This goal means if some partitions of a topic are already
+    * in OfflinePartition state when deletion starts, we need to change the corresponding partition
+    * states to NonExistentPartition first before starting the deletion.
+    *
+    * However we can NOT change partition states to NonExistentPartition at the time of enqueuing topics
+    * for deletion. The reason is that when a topic is enqueued for deletion, it may be ineligible for
+    * deletion due to ongoing partition reassignments. Hence there might be a delay between enqueuing
+    * a topic for deletion and the actual start of deletion. In this delayed interval, partitions may still
+    * transition to or out of the OfflinePartition state.
+    *
+    * Hence we decide to change partition states to NonExistentPartition only when the actual deletion have started.
+    * For topics whose deletion have actually started, we keep track of them in the following topicsBeingDeleted
+    * variable. And once a topic is in the topicsBeingDeleted set, we are sure there will no longer
+    * be partition reassignments to any of its partitions, and only then it's safe to move its partitions to
+    * NonExistentPartition state. Once a topic is in the topicsBeingDeleted set, we will stop monitoring
+    * its partition state changes in the offlinePartitionCount metric
+    */
+  val topicsBeingDeleted = mutable.Set.empty[String]
```

Comment:
Is topicsBeingDeleted the same as topicsToBeDeleted -- topicsIneligibleForDeletion?

Reply:
Not necessarily. While thinking about this problem, instead of topicsBeingDeleted, I think a better name is topicsWithDeletionStarted.

In the following sequence of events, the formula above does not always hold:
1. topic "A" enqueued for deletion
topicsToBeDeleted {A}
topicsWithDeletionStarted {}
topicsIneligibleForDeletion {}

2. starting topic deletion
topicsToBeDeleted {A}
topicsWithDeletionStarted {A}
topicsIneligibleForDeletion {}

3. While processing TopicDeletionStopReplicaResponseReceived, topic "A" is marked ineligible for deletion because the stopping replica failed
topicsToBeDeleted {A}
topicsWithDeletionStarted {A}
topicsIneligibleForDeletion {A}

4. A broker gets restarted, and it has replicas to be deleted:
topicsToBeDeleted {A}
topicsWithDeletionStarted {A}
topicsIneligibleForDeletion {}
...

Reply:
This makes sense.